internal/charger/tplink.go

```diff
@@ -0,0 +1,194 @@
+	}
+	
+	return NewTPLink(cc.URI, cc.StandbyPower)
+}
+
+// NewTPLink creates TP-Link charger
+func NewTPLink(uri string, standbypower float64) (*TPLink, error) {
+	c := &TPLink{
+		uri:          strings.TrimRight(uri, "/"),
+		standbypower: standbypower,
+	}
+	return c, nil
+}
+
+// Enabled implements the Charger.Enabled interface
+func (c *TPLink) Enabled() (bool, error) {
+	sysResp, err := c.execCmd(`{ "system":{ "get_sysinfo":null } }`)
+	if err != nil {
+		return false, err
+	}
+	
+	var systemResponse tplink.SystemResponse
+	if err := json.Unmarshal(sysResp, &systemResponse); err != nil {
+		return false, err
+	}
+	
+	if err := systemResponse.System.GetSysinfo.ErrCode; err != 0 {
+		return false, fmt.Errorf("get_sysinfo error %d", err)
+	}
+	
+	if !strings.Contains(systemResponse.System.GetSysinfo.Feature, "ENE") {
+		return false, errors.New(systemResponse.System.GetSysinfo.Model + " not supported, energy meter feature missing")
+	}
+	
+	return int(1) == systemResponse.System.GetSysinfo.RelayState, err
+}
+
+// Enable implements the Charger.Enable interface
+func (c *TPLink) Enable(enable bool) error {
+	cmd := `{"system":{"set_relay_state":{"state":0}}}`
+	if enable {
+		cmd = `{"system":{"set_relay_state":{"state":1}}}`
+	}
+
+	// Execute TP-Link set_relay_state command
+	sysResp, err := c.execCmd(cmd)
+	if err != nil {
+		return err
+	}
+	
+	var systemResponse tplink.SystemResponse
+	if err := json.Unmarshal(sysResp, &systemResponse); err != nil {
+		return err
+	}
+	
+	if err := systemResponse.System.SetRelayState.ErrCode; err != 0 {
+		return fmt.Errorf("set_relay_state error %d", err)
+	}
+	
+	return nil
+}
+
+// MaxCurrent implements the Charger.MaxCurrent interface
+func (c *TPLink) MaxCurrent(current int64) error {
+	return nil
+}
+
+// Status implements the Charger.Status interface
+func (c *TPLink) Status() (api.ChargeStatus, error) {
+	power, err := c.CurrentPower()
+
+	switch {
+	case power > 0:
+		return api.StatusC, err
+	default:
+		return api.StatusB, err
+	}
+}
+
+var _ api.Meter = (*TPLink)(nil)
+
+// CurrentPower implements the api.Meter interface
+func (c *TPLink) CurrentPower() (float64, error) {
+	emeResp, err := c.execCmd(`{ "emeter":{ "get_realtime":null } }`)
+	if err != nil {
+		return 0, err
+	}
+	
+	var emeterResponse tplink.EmeterResponse
+	if err := json.Unmarshal(emeResp, &emeterResponse); err != nil {
+		return 0, err
+	}
+	
+	if err := emeterResponse.Emeter.GetRealtime.ErrCode; err != 0 {
+		return 0, errors.New("get_realtime error %d", err)
+	}
+
+	power := emeterResponse.Emeter.GetRealtime.Power
+
+	// ignore standby power
+	if power < c.standbypower {
+		power = 0
+	}
+
+	return power, err
+}
+
+// execCmd executes an TP-Link Smart Home Protocol command and provides the response
+func (c *TPLink) execCmd(cmd string) ([]byte, error) {
+	hexHeader := []byte{0, 0, 0, 0} // BigEndian, unsigned integer
+
+	// encode command message
+	// encResult provides the encrypted plug command
+	encCommand := hexHeader
```

Comment:
Warum nicht einfach 

    encCommand := []byte{0, 0, 0, 0}

was hat das mit bigendian zu tun? Das ist einfach eine 0?

Reply:
Done

Reply:
> was hat das mit bigendian zu tun?

Die Phyton encrypt Funktion (siehe unten) erreicht das Ganze durch das entsprechende pack Format.
Bin davon ausgegangen, dass dass irgendwie ein Standard ist, den es auch in go gibt ... sieht halt schöner aus, als platt einfach die 4 x00 davor zu setzen. :-)

def encrypt(string):
    key = 171
    result = **pack(">I",** len(string))
    for i in string:
        a = key ^ ord(i)
        key = a
        result += bytes([a])
    return result